awkward/array/jagged.py

```diff
@@ -1350,9 +1362,42 @@ def concatenate(isclassmethod, cls_or_self, arrays):
 
         return cls(starts, stops, content)
 
-    @classmethod
-    def stack(cls, first, *rest):    # each item in first followed by second, etc.
-        raise NotImplementedError
+    @awkward.util.bothmethod
+    def stack(isclassmethod, cls_or_self, arrays):    # each item in first followed by second, etc.
+
+        if isclassmethod: 
+            cls = cls_or_self
+            if not all(isinstance(x, JaggedArray) for x in arrays):
+                raise TypeError("cannot stack non-JaggedArrays with JaggedArray.stack")
+        else:
+            self = cls_or_self
+            cls = self.__class__
+            if not isinstance(self, JaggedArray) or not all(isinstance(x, JaggedArray) for x in arrays):
+                raise TypeError("cannot stack non-JaggedArrays with JaggedArray.stack")
+            arrays = (self,) + tuple(arrays)
+
+        if len(set([len(a) for a in arrays])) > 1:
+            raise ValueError("cannot stack JaggedArrays of different lengths")
+
+        np = cls.numpy
+
+        n = len(arrays)
+        n_content = sum([len(a.flatten()) for a in arrays])
+        dtype = arrays[0].content.dtype
+
+        # the first step is to get the starts and stops for the stacked structure
+        counts = np.vstack([a.counts for a in arrays])
+        flat_counts = counts.T.flatten()
+        offsets = cls.counts2offsets(flat_counts)
+        starts, stops = offsets[:-1], offsets[1:]
+
+        # use masks for each of the arrays so we can fill the stacked content array at the right indices
+        content = np.zeros(n_content, dtype=dtype)
+        for i in range(n):
+            mask = cls.startsstops2mask(starts[i::n], stops[i::n], length=n_content)
+            content[mask] = arrays[i].flatten()
+
+        return cls(starts[::n], stops[n-1::n], content)
```

Comment:
This is very general and it's only missing one (rarely used) corner case: `starts` and `stops` are allowed to be multidimensional. If `starts` and `stops` have a shape like `(3, 5, 4, million)`, that means that the jagged array has three fixed dimensions—3, 5, and 4—followed by the jagged dimension, and then any fixed dimensions or whatever are inside the `content`.

Ah, there's another case (not so corner): if the `content` of the `arrays` are not `numpy.ndarray`, then this will fail. For instance, the `content` could be `MaskedArrays` or (very likely) `ObjectArrays` (think TLorentzVectorArray). Those cases will require some care to handle: if the `ObjectArray` contains `numpy.ndarray`, concatenating the interiors and re-wrapping them is the best choice; otherwise, the best choice is probably to make the content a `UnionArray`.

The good news is that you don't have to handle them: just check for them and `raise NotImplementedError` if they're not handled. The stub will be a reminder that these cases exist, which has two benefits: (1) if a user encounters them, they'll be informed that it's not something they did wrong; it's not implemented, and (2) it's a hint to future developers that the case is waiting.

Reply:
Ok! The checks are a bit further up in lines 1366 and 1374.

Reply:
Thanks for being explicit about the NotImplementedErrors. At least there won't be confusion.